pymchelper/averaging.py

```diff
@@ -0,0 +1,214 @@
+"""
+Output from the simulation running on multiple processess needs to be agregated.
+The simplest way of doing so is to calculated the average of the data.
+There are however more sophisticated cases: COUNT scorers needs to be summed up, not averaged.
+Phase space data needs to be concatenated, not averaged.
+Each of the parallel jobs can have different number of histories, so the weighted average needs to be calculated.
+Moreover, in case averaging is employed, we can estimate spread of the data as standard deviation or standard error.
+
+The binary output files from each job may be quite large (even ~GB for scoring in fine 3D mesh), to obtain good
+statistics we sometimes parallelise the simulation of hundreds or thousands of jobs (when using HPC clusters).
+In such case it's not feasible to load all the data into memory and calculate the average in one go, using standard
+functions from numpy library. Instead, we need to calculate the average in an online manner,
+i.e. by updating the state of respective aggregator object with each new binary output file read.
+
+Such approach results in a significant reduction of memory usage and is more numerically stable.
+
+This module contains several classes for aggregating data from multiple files:
+- Aggregator: base class for all other aggregators
+- WeightedStatsAggregator: for calculating weighted (using reliability, not frequency weights) mean and variance
+- ConcatenatingAggregator: for concatenating data
+- SumAggregator: for calculating sum instead of variance
+- NoAggregator: for cases when no aggregation is required
```

Comment:
WeightedStatsAggregator  = SH_POSTPROC_AVER 
ConcatenatingAggregator = SH_POSTPROC_APPEND 
SumAggregator = SH_POSTPROC_SUM
NoAggregator = SH_POSTPROC_NONE

how about case SH_POSTPROC_NORM, which would be count/primary ?

Reply:
The code in `averaging.py`  is SHIELD-HIT12A agnostic. It just calculates various types of aggregators. 
The `SH_POSTPROC` numbers are not used in this file, to see how it is used in context of SHIELD-HIT12A take a look at `input_output.py`

`SH_POSTPROC_*` corresponds to https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/input_output.py#L22-L29 

also take a look at:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/input_output.py#L119-L120

the `page_normalized` is a `page` feature extracted from BDO file.

In some cases (for example normalization = 2) the BDO binary file contain number which is not normalized per-primary. Then such normalization happens in the `fromfile` method of the pymchelper reader:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/readers/shieldhit/reader_bdo2019.py#L165-L171

This way estimators obtained using `fromfile` method is already normalized "per-particle".  That holds for all cases: TLET/DLET and Dose/Fluence and others.

With such assumptions its OK to apply weighted status for `AveragingCumulative` (2) and `AveragingPerPrimary` (3)

That fact needs a comment in the code I think !

Reply:
Just FYI, the SH_POSTPROC_SUM was not supposed to be per primary, even after processing the jobs. It is indeed a total count, I guess useful for debugging

Reply:
Implemented in set of last 6 commits, see https://github.com/DataMedSci/pymchelper/pull/723/files/93cfda176bce289504c2a72bb1122166f6e080ac..52947dc09b1cc2af172213102541f243ca132095